Scikit-learn build failed

## Issue description
I am trying to build python36 interpreter with scikit-learn package in it, but it fails with this error: 
```
.................................................................
======================================================================
FAIL: sklearn.linear_model.tests.test_logistic.test_max_iter
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/nix/store/8wbvq1w96h4mmgrn0nxixlkmjd84biii-python3.6-nose-1.3.7/lib/python3.6/site-packages/nose/case.py", line 197, in runTest
    self.test(*self.arg)
  File "/nix/store/a4hmi7338y3738lyklvps2ab687wypdm-python3.6-scikit-learn-0.19.1/lib/python3.6/site-packages/sklearn/utils/testing.py", line 291, in wrapper                                                                                                                             
    return fn(*args, **kwargs)
  File "/nix/store/a4hmi7338y3738lyklvps2ab687wypdm-python3.6-scikit-learn-0.19.1/lib/python3.6/site-packages/sklearn/linear_model/tests/test_logistic.py", line 1023, in test_max_iter                                                                                                   
    assert_equal(lr.n_iter_[0], max_iter)
AssertionError: 0 != 1
    '0 != 1' = '%s != %s' % _common_shorten_repr(0, 1)
    '0 != 1' = self._formatMessage('0 != 1', '0 != 1')
>>  raise self.failureException('0 != 1')


----------------------------------------------------------------------
Ran 8229 tests in 170.089s

FAILED (SKIP=62, failures=1)
builder for '/nix/store/j6q7i7vaynv4cfpq7rwqr3qng8w6yiw3-python3.6-scikit-learn-0.19.1.drv' failed with exit code 1
cannot build derivation '/nix/store/b8zf7b5hd3gzzc9xwrnn0pji5bz9rdi2-python3-3.6.5-env.drv': 1 dependencies couldn't be built
error: build of '/nix/store/b8zf7b5hd3gzzc9xwrnn0pji5bz9rdi2-python3-3.6.5-env.drv' failed
```

### Steps to reproduce
My default.nix:
```
with import <nixpkgs> {};

(let
  python = let
    packageOverrides = self: super: rec {
    };
  in pkgs.python3.override {inherit packageOverrides;};

in python.withPackages(ps: with ps; [
  scikitlearn
])).env
```

Just run `nix-shell` in dir with default.nix

## Technical details

Please run `nix-shell -p nix-info --run "nix-info -m"` and paste the
results.

- system: `"x86_64-linux"`
 - host os: `Linux 4.4.0-130-generic, Ubuntu, 16.04.4 LTS (Xenial Xerus)`
 - multi-user?: `no`
 - sandbox: `no`
 - version: `nix-env (Nix) 2.0.4`
 - channels(user): `"nixpkgs-18.09pre144939.14a9ca27e69"`
 - nixpkgs: `/home/user/.nix-defexpr/channels/nixpkgs`

--- a/pkgs/development/python-modules/scikitlearn/default.nix
+++ b/pkgs/development/python-modules/scikitlearn/default.nix
@@ -14,6 +14,10 @@ buildPythonPackage rec {
     sha256 = "5ca0ad32ee04abe0d4ba02c8d89d501b4e5e0304bdf4d45c2e9875a735b323a0";
   };
 
+  # basically https://github.com/scikit-learn/scikit-learn/pull/10723,
+  # but rebased onto 0.19.1
+  patches = [ ./n_iter-should-be-less-than-max_iter-using-lbgfs.patch ];
+
   buildInputs = [ nose pillow gfortran glibcLocales ];
   propagatedBuildInputs = [ numpy scipy numpy.blas ];
 

--- a/pkgs/development/python-modules/scikitlearn/n_iter-should-be-less-than-max_iter-using-lbgfs.patch
+++ b/pkgs/development/python-modules/scikitlearn/n_iter-should-be-less-than-max_iter-using-lbgfs.patch
@@ -1,0 +1,73 @@
+diff --git a/sklearn/linear_model/huber.py b/sklearn/linear_model/huber.py
+index e17dc1e..665654d 100644
+--- a/sklearn/linear_model/huber.py
++++ b/sklearn/linear_model/huber.py
+@@ -181,7 +181,11 @@ class HuberRegressor(LinearModel, RegressorMixin, BaseEstimator):
+ 
+     n_iter_ : int
+         Number of iterations that fmin_l_bfgs_b has run for.
+-        Not available if SciPy version is 0.9 and below.
++
++        .. versionchanged:: 0.20
++
++            In SciPy <= 1.0.0 the number of lbfgs iterations may exceed
++            ``max_iter``. ``n_iter_`` will now report at most ``max_iter``.
+ 
+     outliers_ : array, shape (n_samples,)
+         A boolean mask which is set to True where the samples are identified
+@@ -272,7 +276,9 @@ class HuberRegressor(LinearModel, RegressorMixin, BaseEstimator):
+             raise ValueError("HuberRegressor convergence failed:"
+                              " l-BFGS-b solver terminated with %s"
+                              % dict_['task'].decode('ascii'))
+-        self.n_iter_ = dict_.get('nit', None)
++        # In scipy <= 1.0.0, nit may exceed maxiter.
++        # See https://github.com/scipy/scipy/issues/7854.
++        self.n_iter_ = min(dict_.get('nit', None), self.max_iter)
+         self.scale_ = parameters[-1]
+         if self.fit_intercept:
+             self.intercept_ = parameters[-2]
+diff --git a/sklearn/linear_model/logistic.py b/sklearn/linear_model/logistic.py
+index 8646c9a..c72a7d9 100644
+--- a/sklearn/linear_model/logistic.py
++++ b/sklearn/linear_model/logistic.py
+@@ -718,7 +718,9 @@ def logistic_regression_path(X, y, pos_class=None, Cs=10, fit_intercept=True,
+                 warnings.warn("lbfgs failed to converge. Increase the number "
+                               "of iterations.")
+             try:
+-                n_iter_i = info['nit'] - 1
++                # In scipy <= 1.0.0, nit may exceed maxiter.
++                # See https://github.com/scipy/scipy/issues/7854.
++                n_iter_i = min(info['nit'], max_iter)
+             except:
+                 n_iter_i = info['funcalls'] - 1
+         elif solver == 'newton-cg':
+@@ -1115,6 +1117,11 @@ class LogisticRegression(BaseEstimator, LinearClassifierMixin,
+         it returns only 1 element. For liblinear solver, only the maximum
+         number of iteration across all classes is given.
+ 
++        .. versionchanged:: 0.20
++
++            In SciPy <= 1.0.0 the number of lbfgs iterations may exceed
++            ``max_iter``. ``n_iter_`` will now report at most ``max_iter``.
++
+     See also
+     --------
+     SGDClassifier : incrementally trained logistic regression (when given
+diff --git a/sklearn/linear_model/tests/test_huber.py b/sklearn/linear_model/tests/test_huber.py
+index 08f4fdf..ca1092f 100644
+--- a/sklearn/linear_model/tests/test_huber.py
++++ b/sklearn/linear_model/tests/test_huber.py
+@@ -42,6 +42,13 @@ def test_huber_equals_lr_for_high_epsilon():
+     assert_almost_equal(huber.intercept_, lr.intercept_, 2)
+ 
+ 
++def test_huber_max_iter():
++    X, y = make_regression_with_outliers()
++    huber = HuberRegressor(max_iter=1)
++    huber.fit(X, y)
++    assert huber.n_iter_ == huber.max_iter
++
++
+ def test_huber_gradient():
+     # Test that the gradient calculated by _huber_loss_and_gradient is correct
+     rng = np.random.RandomState(1)

